### PR TITLE
Ignore HOMEBREW_NO_CLEANUP_FORMULAE deps in `brew autoremove`

### PR DESCRIPTION
**Pull request by apainintheneck**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This is related to #13729 but as @carlocab already mentioned that specific bug might be caused by something else. Either way it got me to reevaluate some of the autoremove logic and I found a bug related to `HOMEBREW_NO_CLEANUP_FORMULAE`.

Essentially what was happening was we were only ignoring the formulae listed in `HOMEBREW_NO_CLEANUP_FORMULAE` but not their dependencies. This PR changes the logic to ignore all of that.

For example, I have `ghostscript` installed. I modified the tab file to indicate that it wasn't installed on request. That would normally make it and its dependencies candidates for `autoremove` as you can see below.

```shell
$ brew autoremove --dry-run
==> Would autoremove 10 unneeded formulae:
fontconfig
freetype
ghostscript
jbig2dec
jpeg-turbo
libidn
libpng
libtiff
little-cms2
openjpeg
```

## Before

When I run the command now with `HOMEBREW_NO_CLEANUP_FORMULAE=ghostscript` it tries to uninstall all of it's dependencies but is luckily stopped by the check in `Uninstall#check_for_dependents`.

```shell
$ HOMEBREW_NO_CLEANUP_FORMULAE=ghostscript brew autoremove
==> Autoremoving 9 unneeded formulae:
fontconfig
freetype
jbig2dec
jpeg-turbo
libidn
libpng
libtiff
little-cms2
openjpeg
Error: Refusing to uninstall /usr/local/Cellar/libpng/1.6.37, /usr/local/Cellar/freetype/2.12.1, /usr/local/Cellar/fontconfig/2.14.0, /usr/local/Cellar/jbig2dec/0.19, /usr/local/Cellar/jpeg-turbo/2.1.4, /usr/local/Cellar/libidn/1.41, /usr/local/Cellar/libtiff/4.4.0_1, /usr/local/Cellar/little-cms2/2.13.1_1 and /usr/local/Cellar/openjpeg/2.5.0
because they are required by ghostscript, which is currently installed.
You can override this and force removal with:
  brew uninstall --ignore-dependencies
```

## After

When I rerun the command now with `HOMEBREW_NO_CLEANUP_FORMULAE=ghostscript` it returns no output which is what we want.

```shell
$ HOMEBREW_NO_CLEANUP_FORMULAE=ghostscript brew autoremove --dry_run
```

### CODE REVIEW

**Review by carlocab — APPROVED** [review]

File: Library/Homebrew/cleanup.rb [hunk 1]
```diff
@@ -535,8 +535,12 @@ def self.autoremove(dry_run: false)
       # the cache of installed formulae may no longer be valid.
       Formula.clear_cache unless dry_run
 
-      # Remove formulae listed in HOMEBREW_NO_CLEANUP_FORMULAE.
-      formulae = Formula.installed.reject(&method(:skip_clean_formula?))
+      formulae = Formula.installed
+      # Remove formulae listed in HOMEBREW_NO_CLEANUP_FORMULAE and their dependencies.
+      if Homebrew::EnvConfig.no_cleanup_formulae.present?
+        formulae -= formulae.select(&method(:skip_clean_formula?))
+                            .flat_map { |f| f.runtime_formula_dependencies + [f] }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cleanup.rb:542]:
```suggestion
                            .flat_map { |f| [f, *f.runtime_formula_dependencies] }
```

**Review by MikeMcQuaid — APPROVED** [review]
Looks good, nice work! Suggested, optional style tweak.

File: Library/Homebrew/cleanup.rb (hunk 1 above)
Comment (apainintheneck) [reply Library/Homebrew/cleanup.rb]: Ah yeah, that looks better.

**Comment by MikeMcQuaid** [comment]
Thanks again @apainintheneck!

**Pull request merged**